src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,345 @@
+        assert isinstance(model, PreTrainedModel)
+        if isinstance(config, dict):
+            assert len(config) == 1
+            peft_config = config[adapter_name]
+        else:
+            peft_config = config
+        assert isinstance(peft_config, xLoRAConfig)
+
+        super().__init__(model, config, adapter_name, model_peft, _disable_inject=True)
+
+        if hasattr(model.config, "use_cache"):
+            assert not model.config.use_cache, "`use_cache` must be False"
+
+        use_trainable_adapters = peft_config.use_trainable_adapters
+        adapters_items = iter(peft_config.adapters.items())
+
+        # Because we call load_adapter, which requires base_model to be defined
+        model_peft.base_model = self
+        # For load_adapter to think we are a LoraModel
+        model_peft.peft_type = PeftType.LORA
+
+        for adapter_name, model_id in adapters_items:
+            model_peft.load_adapter(model_id, adapter_name, is_trainable=use_trainable_adapters)
+
+        self.set_adapter(list(peft_config.adapters.keys()))
+        model_peft.peft_type = PeftType.XLORA
+
+        def hook(module, *args, **kwargs) -> None:
+            args_real = args[0]
+            kwargs_real: dict = args[1]
+            kwargs_real.update(kwargs)
+
+            xlora_classifier: xLoRAClassifier = self.internal_xlora_classifier  # type: ignore
+
+            if "_xlora_classifier_inhibitor_flag" in kwargs_real:
+                payload: InhibitorFlagPayload = kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                del kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                self.internal_xlora_scalings = torch.full(  # type: ignore
+                    (payload.batch_size, payload.seq_len, xlora_classifier.n_layers, xlora_classifier.n_classes),
+                    payload.override_scaling_pass_value,
+                )
+
+                return
+
+            xlora_scalings = xlora_classifier.forward(
+                *args_real,
+                **kwargs_real,
+            )
+            # Set the scalings
+            self.internal_xlora_scalings = xlora_scalings
+
+        model.register_forward_pre_hook(hook, with_kwargs=True, prepend=True)
+
+        self.eval()
+        if not use_trainable_adapters:
+            total_frozen = 0
+            for name, param in self.named_parameters():
+                if "lora_" in name:
+                    param.requires_grad = False
+                    total_frozen += 1
+
+        assert isinstance(self, LoraModel)
+
+        total_swapped = convert_layers_to_xlora(
+            model_peft,
+            peft_config,
+        )
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = xLoRAClassifier(model_peft, peft_config, n_classes, total_swapped)
+
+        # Setup the internal state
+        base_model_wrapper = BaseTunerWrapper(self, xlora_classifier)
+        self.forward = base_model_wrapper.forward  # type: ignore[method-assign]
```

Comment:
Instead of assigning methods like this, could we not define a `def forward` method that calls the appropriate method?

Reply:
That is a much more elegant solution, I will update the code!